### ISSUE TITLE
[linux-headers-surface-6.9.3-surface-1] Issues with launching VMWare Workstation 17

### ISSUE THREAD

**Issue by der-siebte-schatten** [linux-surface/linux-surface#1465]
[Description of the bug or feature]

### Environment

- Hardware model: Surface Book 3
- Kernel version: 6.9.3-surface-1
- Distribution: LMDE 6 "Faye"

Hello there,
I'm experiencing a new issue after updating kernel and headers today: VMWare Workstation 17 won't launch anymore complaining of missing gcc. gcc 12 is installed, even if VMWare displays gcc 9.4.0 because of kernel compiled with this version.

On version 6.8.8-surface-1, I was able to overcome this issue by editing vmnet and vmmon tar files to change some lines, as reported here : https://askubuntu.com/questions/1494239/gnu-c-compiler-gcc-version-12-2-0-was-not-found-vmware. It worked then.
Now this method does not work anymore. Plus, when compiling both drivers manually, I encountered new errors, meaning some changes in the headers are not compatible

<details><summary>`make` output for vmnet</summary>

```
Using kernel build system.
make -C /lib/modules/6.9.3-surface-1/build/include/.. M=$PWD SRCROOT=$PWD/. \
  MODULEBUILDDIR= modules
make[1] : on entre dans le répertoire « /usr/src/linux-headers-6.9.3-surface-1 »
warning: the compiler differs from the one used to build the kernel
  The kernel was built by: gcc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
  You are using:           gcc (Debian 12.2.0-14) 12.2.0
  CC [M]  /usr/lib/vmware/modules/source/vmnet-only/bridge.o
/usr/lib/vmware/modules/source/vmnet-only/bridge.c:256:1: warning: no previous prototype for ‘VNetBridge_Create’ [-Wmissing-prototypes]
  256 | VNetBridge_Create(const char *devName, // IN:  name of device (e.g., "eth0")
      | ^~~~~~~~~~~~~~~~~
In file included from ./include/linux/spinlock.h:305,
                 from ./include/linux/sched.h:2138,
                 from /usr/lib/vmware/modules/source/vmnet-only/bridge.c:25:
/usr/lib/vmware/modules/source/vmnet-only/bridge.c: In function ‘VNetBridgeReceiveFromVNet’:
/usr/lib/vmware/modules/source/vmnet-only/vmnetInt.h:44:39: error: ‘dev_base_lock’ undeclared (first use in this function); did you mean ‘device_lock’?
   44 | #define dev_lock_list()    read_lock(&dev_base_lock)
      |                                       ^~~~~~~~~~~~~
./include/linux/rwlock.h:56:48: note: in definition of macro ‘read_lock’
   56 | #define read_lock(lock)         _raw_read_lock(lock)
      |                                                ^~~~
/usr/lib/vmware/modules/source/vmnet-only/bridge.c:585:4: note: in expansion of macro ‘dev_lock_list’
  585 |    dev_lock_list();
      |    ^~~~~~~~~~~~~
/usr/lib/vmware/modules/source/vmnet-only/vmnetInt.h:44:39: note: each undeclared identifier is reported only once for each function it appears in
   44 | #define dev_lock_list()    read_lock(&dev_base_lock)
      |                                       ^~~~~~~~~~~~~
./include/linux/rwlock.h:56:48: note: in definition of macro ‘read_lock’
   56 | #define read_lock(lock)         _raw_read_lock(lock)
      |                                                ^~~~
/usr/lib/vmware/modules/source/vmnet-only/bridge.c:585:4: note: in expansion of macro ‘dev_lock_list’
  585 |    dev_lock_list();
      |    ^~~~~~~~~~~~~
/usr/lib/vmware/modules/source/vmnet-only/bridge.c: In function ‘VNetBridgeUp’:
/usr/lib/vmware/modules/source/vmnet-only/vmnetInt.h:44:39: error: ‘dev_base_lock’ undeclared (first use in this function); did you mean ‘device_lock’?
   44 | #define dev_lock_list()    read_lock(&dev_base_lock)
      |                                       ^~~~~~~~~~~~~
./include/linux/rwlock.h:56:48: note: in definition of macro ‘read_lock’
   56 | #define read_lock(lock)         _raw_read_lock(lock)
      |                                                ^~~~
/usr/lib/vmware/modules/source/vmnet-only/bridge.c:898:4: note: in expansion of macro ‘dev_lock_list’
  898 |    dev_lock_list();
      |    ^~~~~~~~~~~~~
/usr/lib/vmware/modules/source/vmnet-only/bridge.c: At top level:
/usr/lib/vmware/modules/source/vmnet-only/bridge.c:1409:1: warning: no previous prototype for ‘VNetBridgeSendLargePacket’ [-Wmissing-prototypes]
 1409 | VNetBridgeSendLargePacket(struct sk_buff *skb,        // IN: packet to split
      | ^~~~~~~~~~~~~~~~~~~~~~~~~
make[3]: *** [scripts/Makefile.build:244 : /usr/lib/vmware/modules/source/vmnet-only/bridge.o] Erreur 1
make[2]: *** [/usr/src/linux-headers-6.9.3-surface-1/Makefile:1925 : /usr/lib/vmware/modules/source/vmnet-only] Erreur 2
make[1]: *** [Makefile:240 : __sub-make] Erreur 2
make[1] : on quitte le répertoire « /usr/src/linux-headers-6.9.3-surface-1 »
make: *** [Makefile:117 : vmnet.ko] Erreur 2

```
</details>

<details><summary>`make` output for vmmon</summary>
```
sing kernel build system.
make -C /lib/modules/6.9.3-surface-1/build/include/.. M=$PWD SRCROOT=$PWD/. \
  MODULEBUILDDIR= modules
make[1] : on entre dans le répertoire « /usr/src/linux-headers-6.9.3-surface-1 »
warning: the compiler differs from the one used to build the kernel
  The kernel was built by: gcc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
  You are using:           gcc (Debian 12.2.0-14) 12.2.0
  CC [M]  /usr/lib/vmware/modules/source/vmmon-only/linux/hostif.o
In file included from /usr/lib/vmware/modules/source/vmmon-only/linux/hostif.c:76:
/usr/lib/vmware/modules/source/vmmon-only/./include/pgtbl.h: In function ‘PgtblVa2MPNLocked’:
/usr/lib/vmware/modules/source/vmmon-only/./include/pgtbl.h:60:8: error: implicit declaration of function ‘pgd_large’; did you mean ‘pgd_page’? [-Werror=implicit-function-declaration]
   60 |    if (pgd_large(*pgd)) {
      |        ^~~~~~~~~
      |        pgd_page
In file included from /usr/lib/vmware/modules/source/vmmon-only/./include/pgtbl.h:25:
/usr/lib/vmware/modules/source/vmmon-only/./include/compat_pgtable.h:39:45: error: implicit declaration of function ‘p4d_large’; did you mean ‘p4d_page’? [-Werror=implicit-function-declaration]
   39 | #   define compat_p4d_large(p4d)            p4d_large(p4d)
      |                                             ^~~~~~~~~
/usr/lib/vmware/modules/source/vmmon-only/./include/pgtbl.h:70:8: note: in expansion of macro ‘compat_p4d_large’
   70 |    if (compat_p4d_large(*p4d)) {
      |        ^~~~~~~~~~~~~~~~
/usr/lib/vmware/modules/source/vmmon-only/./include/pgtbl.h:78:11: error: implicit declaration of function ‘pud_large’; did you mean ‘pud_page’? [-Werror=implicit-function-declaration]
   78 |       if (pud_large(*pud)) {
      |           ^~~~~~~~~
      |           pud_page
/usr/lib/vmware/modules/source/vmmon-only/./include/pgtbl.h:86:14: error: implicit declaration of function ‘pmd_large’; did you mean ‘pmd_page’? [-Werror=implicit-function-declaration]
   86 |          if (pmd_large(*pmd)) {
      |              ^~~~~~~~~
      |              pmd_page
/usr/lib/vmware/modules/source/vmmon-only/linux/hostif.c: At top level:
/usr/lib/vmware/modules/source/vmmon-only/linux/hostif.c:2926:1: warning: no previous prototype for ‘HostIFCheckTrackedMPN’ [-Wmissing-prototypes]
 2926 | HostIFCheckTrackedMPN(VMDriver *vm, // IN: The VM instance
      | ^~~~~~~~~~~~~~~~~~~~~
/usr/lib/vmware/modules/source/vmmon-only/linux/hostif.c:3046:1: warning: no previous prototype for ‘HostIFWritePhysicalWork’ [-Wmissing-prototypes]
 3046 | HostIFWritePhysicalWork(MA ma,             // MA to be written to
      | ^~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/vmware/modules/source/vmmon-only/linux/hostif.c:3205:1: warning: no previous prototype for ‘HostIFStartTimer’ [-Wmissing-prototypes]
 3205 | HostIFStartTimer(Bool rateChanged,  //IN: Did rate change?
      | ^~~~~~~~~~~~~~~~
cc1: some warnings being treated as errors
make[3]: *** [scripts/Makefile.build:244 : /usr/lib/vmware/modules/source/vmmon-only/linux/hostif.o] Erreur 1
make[2]: *** [/usr/src/linux-headers-6.9.3-surface-1/Makefile:1925 : /usr/lib/vmware/modules/source/vmmon-only] Erreur 2
make[1]: *** [Makefile:240 : __sub-make] Erreur 2
make[1] : on quitte le répertoire « /usr/src/linux-headers-6.9.3-surface-1 »
make: *** [Makefile:117 : vmmon.ko] Erreur 2
```
</details>

**Comment by der-siebte-schatten** [comment]
So it happens to be common to all 6.9.x Linux kernel, this is not due to linux-surface...

**Comment by qzed** [comment]
This is unfortunately a somewhat common problem. Since we're running a kernel that is on occasion quite a bit newer than what Debian-based distributions ship, the external kernel module sources from that distributions might not be new enough to work with them.
